synapse/handlers/federation.py

```diff
@@ -265,31 +254,99 @@ async def _maybe_backfill_inner(
         #    2. we have likely previously tried and failed to backfill from that
         #       extremity, so to avoid getting "stuck" requesting the same
         #       backfill repeatedly we drop those extremities.
-        filtered_sorted_extremeties_tuple = [
-            t for t in sorted_extremeties_tuple if int(t[1]) <= current_depth
-        ]
-
-        logger.debug(
-            "room_id: %s, backfill: current_depth: %s, limit: %s, max_depth: %s, extrems (%d): %s filtered_sorted_extremeties_tuple: %s",
-            room_id,
-            current_depth,
-            limit,
-            max_depth,
-            len(sorted_extremeties_tuple),
-            sorted_extremeties_tuple,
-            filtered_sorted_extremeties_tuple,
-        )
-
+        #
         # However, we need to check that the filtered extremities are non-empty.
         # If they are empty then either we can a) bail or b) still attempt to
         # backfill. We opt to try backfilling anyway just in case we do get
         # relevant events.
-        if filtered_sorted_extremeties_tuple:
-            sorted_extremeties_tuple = filtered_sorted_extremeties_tuple
+        #
+        filtered_sorted_backfill_points = [
+            t for t in sorted_backfill_points if t.depth <= current_depth
+        ]
+        if filtered_sorted_backfill_points:
+            logger.debug(
+                "_maybe_backfill_inner: backfill points before current depth: %s",
+                filtered_sorted_backfill_points,
+            )
+            sorted_backfill_points = filtered_sorted_backfill_points
+        else:
+            logger.debug(
+                "_maybe_backfill_inner: all backfill points are *after* current depth. Backfilling anyway."
+            )
 
-        # We don't want to specify too many extremities as it causes the backfill
-        # request URI to be too long.
-        extremities = dict(sorted_extremeties_tuple[:5])
+        # We still need to narrow down the list of extremities we pass to the remote
+        # server. We limit to 5 of them, to avoid the request URI becoming too long.
+        #
+        # However, we only want to paginate from a particular extremity if we can
+        # actually see the events we'll get, as otherwise we'll just spend a lot of
+        # resources to get redacted events.
+        #
+        # We do this by filtering all the backwards extremities and seeing if
+        # any remain.
```

Comment:
I still think filtering is an odd verb (even if it makes sense given that we call `filter_events_for_server`). Could this be rewritten? An attempt:

```
        # For performance's sake, we only want to paginate from a particular extremity
        # if we can actually see the events we'll get. Otherwise, we'd just spend a lot
        # of resources to get redacted events. We check each extremity in turn and
        # ignore those which don't merit backfilling.
        # 
        # Additionally, we limit ourselves to backfilling from at most 5 extremities,
        # for two reasons:
        # 
        # - The check which determines if we can see an extremity's events can be
        #   expensive (we load the full state for the room at each of the sucessor 
        #   events).
        # - We want to avoid the server-server API request URI becoming too long.
```

Reply:
(The point in https://github.com/matrix-org/synapse/pull/12522/files#r857783795 notwithstanding)

Reply:
I totally agree it's oddly phrased. It's just something I copied from the original implementation without stopping to think if it could be improved.  I'll update it.

Reply:
done in [8487963](https://github.com/matrix-org/synapse/pull/12522/commits/84879633516be303f5b795cb5aac3468da953d88)